tool/tsh/db.go

```diff
@@ -124,33 +124,42 @@ func databaseLogin(cf *CLIConf, tc *client.TeleportClient, db tlsca.RouteToDatab
 		// ref: https://redis.io/commands/auth
 		db.Username = defaults.DefaultRedisUsername
 	}
-	profile, err := client.StatusCurrent(cf.HomePath, cf.Proxy)
+
+	// Note: we don't use StatusCurrentWithIdentity here as virtual identity
+	// profiles are read-only.
+	profile, err := client.StatusCurrentWithIdentity(cf.HomePath, cf.Proxy, cf.IdentityFileIn)
 	if err != nil {
 		return trace.Wrap(err)
 	}
 
-	var key *client.Key
-	if err = client.RetryWithRelogin(cf.Context, tc, func() error {
-		key, err = tc.IssueUserCertsWithMFA(cf.Context, client.ReissueParams{
-			RouteToCluster: tc.SiteName,
-			RouteToDatabase: proto.RouteToDatabase{
-				ServiceName: db.ServiceName,
-				Protocol:    db.Protocol,
-				Username:    db.Username,
-				Database:    db.Database,
-			},
-			AccessRequests: profile.ActiveRequests.AccessRequests,
-		})
-		return trace.Wrap(err)
-	}); err != nil {
-		return trace.Wrap(err)
-	}
-	if err = tc.LocalAgent().AddDatabaseKey(key); err != nil {
-		return trace.Wrap(err)
+	// Identity files contain themselves act as the database credentials (if
+	// any), so don't bother fetching new certs.
+	if profile.IsVirtual {
```

Comment:
Is there a way to restructure the code so that we don't increase the complexity massively by having each and every operation support "virtual profiles" in explicit way?

Reply:
I don't much like the verbosity either, but ultimately certain codepaths (like this one) do things that will fail when given an identity file. Arguably when using an identity file I don't think any additional certs should get written to disk, and moreover `RetryWithRelogin()` and `IssueUserCertsWithMFA()` will fail on Machine ID identity files as they're explicitly non-renewable.

If we (per your comment below) made `RetryWithRelogin` and friends identity-aware (i.e. effectively a no-op), we might be able to skip more of these checks. It's hard for me to predict if that'd be significantly simpler, though.

(and wow, that comment is unreadable, whoops...)

Reply:
With the latest change I've merged everything into the one codepath (`StatusCurrent()`) and we have a total of 3 `IsVirtual` checks for core + database access.

I know app (and probably k8s) access will need some of their own when we circle back to support them properly, but I don't think this is especially onerous - worst case if some specific handling is missing it fails due to some identity-related error exactly as it would already today.

Overall, though, I think the alternative where we modify `RetryWithRelogin` to support identity files ends up quite a bit more complicated and I'd rather direct that effort to a proper refactoring where hacks like this aren't needed at all.